homeassistant/components/apcupsd/__init__.py

```diff
@@ -1,60 +1,130 @@
+        "has been imported into the UI automatically and can be safely removed "
+        "from your configuration.yaml file"
+    )
+
+    # We only import configs from YAML if it hasn't been imported.
+    if _async_get_imported_entry(hass) is None:
+        # Since the YAML configuration for apcupsd consists of two parts:
+        # apcupsd:
+        #   host: xxx
+        #   port: xxx
+        # sensor:
+        #   - platform: apcupsd
+        #     resource:
+        #       - resource_1
+        #       - resource_2
+        #       - ...
+        # Here at the integration set up we do not have the entire information to be
+        # imported to config flow yet. So we temporarily store the configuration to
+        # hass.data[DOMAIN] under a special entry_id SOURCE_IMPORT (which shouldn't
+        # conflict with other entry ids). Later when the sensor platform setup is called
+        # we gather the resources information and from there we start the actual config
+        # entry imports.
+        hass.data.setdefault(DOMAIN, {})
+        hass.data[DOMAIN][SOURCE_IMPORT] = conf
+
+    return True
 
-    apcups_data = APCUPSdData(host, port)
-    hass.data[DOMAIN] = apcups_data
 
-    # It doesn't really matter why we're not able to get the status, just that
-    # we can't.
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Use config values to set up a function enabling status retrieval."""
+    data_service = APCUPSdData(
+        config_entry.data[CONF_HOST], config_entry.data[CONF_PORT]
+    )
+
+    # It doesn't really matter why we're not able to get the status, just that we can't.
     try:
-        apcups_data.update(no_throttle=True)
+        await hass.async_add_executor_job(data_service.update)
     except Exception:  # pylint: disable=broad-except
         _LOGGER.exception("Failure while testing APCUPSd status retrieval")
         return False
+
+    # Set up option update handler.
+    config_entry.async_on_unload(
+        config_entry.add_update_listener(_async_options_updated)
+    )
+
+    # Store the data service object.
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][config_entry.entry_id] = data_service
+
+    # Forward the config entries to the supported platforms.
+    for platform in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, platform)
+        )
     return True
 
 
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+async def _async_options_updated(hass: HomeAssistant, config_entry: ConfigEntry):
+    """Handle options update."""
+    await hass.config_entries.async_reload(config_entry.entry_id)
+
+
 class APCUPSdData:
```

Comment:
Please rewrite `APCUPSdData` so it doesn't call a library method in the `status` property. We shouldn't do I/O when accessing properties. API calls should be explicit.

Reply:
Since this PR was designed to add config flow, the original code was intentionally left untouched. But since we are doing a lot already, `APCUPSdData` class is simplified and now any I/O is explicit.